website/guides/advanced/custom-aggregations-with-lua.md.erb

```diff
@@ -0,0 +1,284 @@
+---
+
+While Vector's built-in [transforms][docs.reference.transforms] are fast, sometimes they are not expressive
+enough for your needs. In such cases the [`lua`][docs.reference.transforms.lua] transform comes to rescue, letting you
+define custom transformation logic.
+
+This guide walks through various ways of defining an aggregating transform component that takes incoming log events,
+counts them, and emits [`counter`][docs.about.data-model.metric#counter] metrics each 5 seconds.
+
+## Architectural Overview
+
+Lua is an interpreted language embedded into Vector. When a `lua` transform is created, it starts an instance
+of the Lua interpreter. As a consequence, different transforms are isolated and cannot interrupt each other.
+
+The execution model is asynchronous, with two key concepts: _hooks_ and _timer handlers_. Both of them are
+user-defined Lua functions which are called by Vector at certain events.
+
+### Hooks
+
+There are three types of hooks: `init`, `process`, and `shutdown`.
+
+#### The `process` hook
+
+The most important of them is `process`, which is called on each incoming events. It can be defined like this:
+
+```toml
+hooks.process = """
+function (event, emit)
+  -- do something
+end
+"""
+```
+
+It takes a single event and can output one or many of them using the `emit` function provided as the second argument.
+
+For example, the body of the function above could have been
+
+```lua
+evnet.log.my_field = "my value"
+emit(event)
+```
+
+The code above sets field `my_field` to value `"my_value"` and sends the newly created event to the downstream
+components. Read more about event representation [in the reference][docs.reference.transforms.lua#representation-of-events].
+
+#### The `init` hook
+
+The `init` hook is similar to `process` hook, but it is called before the first call of the `process` hook, and thus
+takes no events as its arguments.
+
+Note that it although it is called before the first event, it is called only after the first event is ready to be
+processed. However, one should not rely on this behavior, as it is not guaranteed to not change in the future.
+
+#### The `shutdown` hook
+
+The `shutdown` hook is called after the last event is received. It doesn't take events as arguments as well.
+
+### Timer handlers
+
+Timer handlers are similar to hooks by being Lua functions capable of producing events. However, they are called
+periodically at pre-defined intervals.
+
+
+All of the functions listed above share a single runtime, so they can communicate between each other using global
+variables.
+
+## First Implementation
+
+Using the knowledge from the previous section, it is possible to write down the following transform definition:
+
+```toml title="vector.toml"
+[transforms.aggregator]
+type = "lua"
+version = "2"
+inputs = [] # add names of the input components here
+
+hooks.init = """
+  function (emit)
+    count = 0 -- initialize state by setting a global variable
```

Comment:
One thing I noticed, is this an actual global variable since it is contained within a function?

Reply:
Yes, it is global. Local variables are prefixed by `local` keyword. It is similar to JavaScript, where assigning a variable without `var`/`let`/`const` makes it global.